File: Go/jobspls/algorithms/perm_missing_element.go
package algorithms

func PermMissingElement(haystack []int) uint {
	hayset := make(map[int]struct{}, len(haystack))

	for _, v := range haystack {
		hayset[v] = struct{}{}
	}

	// Will always return at least element because the loop goes to 
	// one past haystack's length 
	for i := 1; i < len(haystack) + 2; i++ {
		if _, found := hayset[i]; !found {
			return uint(i)
		}
	}

	panic("Unreachable")
}

File: Java/jobspls/algorithms/PermMissingElement.java
package jobspls.algorithms;

import java.util.Arrays;
import java.util.TreeSet;

import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class PermMissingElement {
    public static int missing_element(int[] haystack) {
        var hayset = new TreeSet<Integer>(Arrays.stream(haystack)
                                                .boxed()
                                                .collect(Collectors.toList()));

        // Should always find a value.
        return IntStream.range(1, haystack.length + 2)
                        .filter(n -> { return !hayset.contains(n); })
                        .findFirst()
                        .getAsInt();
    }
}
